pkg/resource/rule/hooks.go

```diff
@@ -0,0 +1,144 @@
+package rule
+
+import (
+	"context"
+	"fmt"
+
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
+	"github.com/aws/aws-sdk-go/aws"
+	svcsdk "github.com/aws/aws-sdk-go/service/eventbridge"
+
+	"github.com/aws-controllers-k8s/eventbridge-controller/apis/v1alpha1"
+	svcapitypes "github.com/aws-controllers-k8s/eventbridge-controller/apis/v1alpha1"
+)
+
+type validationError struct {
+	field   string
+	message string
+}
+
+func (v validationError) Error() string {
+	return fmt.Sprintf("invalid Spec: %q: %s", v.field, v.message)
+}
+
+func newValidationError(field, message string) validationError {
+	return validationError{
+		field:   field,
+		message: message,
+	}
+}
+
+func validateRuleSpec(spec v1alpha1.RuleSpec) error {
+	var match bool
+	if s := spec.State; s != nil {
+		allowedValues := svcsdk.RuleState_Values()
+		for _, v := range allowedValues {
+			if *s == v {
+				match = true
+			}
+		}
+		if !match {
+			return newValidationError(
+				"spec.state",
+				fmt.Sprintf("supported states: %v", allowedValues),
+			)
+		}
+	}
+
+	emptyPattern := func() bool {
+		return spec.EventPattern == nil || *spec.EventPattern == ""
+	}
+
+	emptySchedule := func() bool {
+		return spec.ScheduleExpression == nil || *spec.ScheduleExpression == ""
+	}
```

Comment:
Any particular reason these are lambda functions and not just boolean literals?

Reply:
I created those anonymous functions to make the code in line 57 more readable:

`if emptySchedule() && emptyPattern() {...}`

How would you rewrite this code? Just curious as I don't quiet understand your ask here :)

Reply:
Just like this?
```
emptySchedule := (spec.ScheduleExpression == nil || *spec.ScheduleExpression == "")
```

Reply:
Yeah, my bad 🤦🏻‍♂️